core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala

```diff
@@ -0,0 +1,267 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * 
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.api
+
+import java.io.File
+import java.util.ArrayList
+import java.util.concurrent.ExecutionException
+
+import kafka.admin.AclCommand
+import kafka.common.TopicAndPartition
+import kafka.security.auth._
+import kafka.server._
+import kafka.utils._
+
+import org.apache.kafka.clients.consumer.{Consumer, ConsumerRecord, ConsumerConfig}
+import org.apache.kafka.clients.producer.{ProducerRecord, ProducerConfig}
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+import org.apache.kafka.common.{TopicPartition}
+import org.apache.kafka.common.protocol.SecurityProtocol
+import org.apache.kafka.common.errors.{GroupAuthorizationException,TopicAuthorizationException}
+import org.junit.Assert._
+import org.junit.{Test, After, Before}
+
+import scala.collection.JavaConverters._
+
+
+/**
+  * The test cases here verify that a producer authorized to publish to a topic
+  * is able to, and that consumers in a group authorized to consume are able to
+  * to do so.
+  *
+  * This test relies on a chain of test harness traits to set up. It directly
+  * extends IntegrationTestHarness. IntegrationTestHarness creates producers and
+  * consumers, and it extends KafkaServerTestHarness. KafkaServerTestHarness starts
+  * brokers, but first it initializes a ZooKeeper server and client, which happens
+  * in ZooKeeperTestHarness.
+  *
+  * To start brokers when the security protocol is SASL_SSL, we need to set a cluster
+  * ACL, which happens optionally in KafkaServerTestHarness. If the security protocol
+  * is SSL or PLAINTEXT, then the ACL isn't set. The remaining ACLs to enable access
+  * to producers and consumers are set here. To set ACLs, we use AclCommand directly.
+  *
+  * Finally, we rely on SaslSetup to bootstrap and setup Kerberos. We don't use
+  * SaslTestHarness here directly because it extends ZooKeeperTestHarness, and we
+  * would end up with ZooKeeperTestHarness twice.
+  */
+trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+  override val producerCount = 1
+  override val consumerCount = 2
+  override val serverCount = 3
+  override val setClusterAcl = Some(() => AclCommand.main(clusterAclArgs))
```

Comment:
We will need to call TestUtils.waitAndVerifyAcls here too. To do that, we probably need to move the logic of calling setClusterAcl to the end of KafkaServerTestHarness.setup. Also, could we just move the setClusterAcl logic to EndToEndAuthorizationTest.setup()?